### PR TITLE
APS-1800 Updated amend booking to change room characteristics.

### PR DESCRIPTION
**Pull request by vbala-moj**
[https://dsdmoj.atlassian.net/browse/APS-1800](url)

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1BookingDomainEventService.kt [hunk 1]
```diff
@@ -524,5 +531,13 @@ class Cas1BookingDomainEventService(
     val previousArrivalDateIfChanged: LocalDate?,
     val previousDepartureDateIfChanged: LocalDate?,
     val isSpaceBooking: Boolean,
+    val characteristics: List<Cas1SpaceCharacteristic> = emptyList(),
+    val previousCharacteristics: List<Cas1SpaceCharacteristic>? = emptyList(),
```
Comment (vbala-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1BookingDomainEventService.kt:535]: Not sure if we need this for the old booking, so default to emptyList() so not affect the bookingChanged() function.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/cas1/Cas1SpaceBookingController.kt [hunk 2]
```diff
@@ -168,13 +168,23 @@ class Cas1SpaceBookingController(
 
     val user = userService.getUserForRequest()
 
+    val characteristics = cas1UpdateSpaceBooking.characteristics
+      ?.takeIf { it.isNotEmpty() }
```
Comment (davidatkinsuk) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/cas1/Cas1SpaceBookingController.kt:172]: i'm not sure the `.takeIf` is necessary. I think if the list is empty all the following functions will essential be skipped over?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/cas1/Cas1SpaceBookingController.kt [hunk 3]
```diff
@@ -168,13 +168,23 @@ class Cas1SpaceBookingController(
...
+      ?.map { it.value }
+      ?.let { values ->
+        characteristicService.getCharacteristicsByPropertyNames(values, ServiceName.approvedPremises)
+      }
+      ?.filter { it.isModelScopeRoom() }
+      ?: emptyList()
```
Comment (davidatkinsuk) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/cas1/Cas1SpaceBookingController.kt:178]: it may be a little cleaner if you move `?: emptyList()` to the first line. We can then remove all the `?`s  e.g.

```
val characteristics = (cas1UpdateSpaceBooking.characteristics ?: emptyList())
      .map { it.value }
      .let { values ->
        characteristicService.getCharacteristicsByPropertyNames(values, ServiceName.approvedPremises)
      }
      .filter { it.isModelScopeRoom() }
```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/domainevents/DomainEventDescriber.kt [hunk 4]
```diff
@@ -137,6 +138,10 @@ class DomainEventDescriber(
         addDateChangeMessage(previousDeparture, eventDetails.departureOn, "departure")
       }
 
+      if (!previousCharacteristics.isNullOrEmpty()) {
+        changes.add("its booking amended")
+      }
```
Comment (davidatkinsuk) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/domainevents/DomainEventDescriber.kt:143]: given the discussion yesterday i think we should remove these new lines as we'll be shifting this logic into the UI

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/CharacteristicEntity.kt [hunk 5]
```diff
@@ -95,4 +95,5 @@ data class CharacteristicEntity(
   fun matches(entityServiceScope: String, entityModelScope: String) = serviceMatches(entityServiceScope) && modelMatches(entityModelScope)
   fun serviceMatches(entityServiceScope: String) = serviceScope == "*" || entityServiceScope == serviceScope
   fun modelMatches(entityModelScope: String) = modelScope == "*" || entityModelScope == modelScope
+  fun isModelScopePremises() = modelMatches("premises")
```
Comment (davidatkinsuk) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/CharacteristicEntity.kt:98]: nice

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1SpaceBookingService.kt [hunk 6]
```diff
@@ -567,6 +568,7 @@ class Cas1SpaceBookingService(
       bookingChangedAt = OffsetDateTime.now(),
       previousArrivalDateIfChanged = if (previousArrivalDate != updatedBooking.expectedArrivalDate) previousArrivalDate else null,
       previousDepartureDateIfChanged = if (previousDepartureDate != updatedBooking.expectedDepartureDate) previousDepartureDate else null,
+      previousCharacteristicsIfChanged = if (previousCharacteristics != updatedBooking.criteria) previousCharacteristics else emptyList(),
```
Comment (davidatkinsuk) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1SpaceBookingService.kt:571]: will this work if the characteristics are in different order in the two lists? i think default should be null in this case (As discussed in a ticket elsewhere). emptyList() implies that there were no characteristics previously, which may not be the case. Can we unit test this?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1SpaceBookingService.kt [hunk 7]
```diff
@@ -615,9 +617,23 @@ class Cas1SpaceBookingService(
...
+
     return cas1SpaceBookingRepository.save(bookingToUpdate)
   }
 
+  private fun updateRoomCharacteristics(
+    booking: Cas1SpaceBookingEntity,
+    newRoomCharacteristics: List<CharacteristicEntity>,
+  ) {
+    booking.criteria.apply {
+      retainAll { it.isModelScopePremises() }
```
Comment (davidatkinsuk) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1SpaceBookingService.kt:632]: nice!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1SpaceBookingService.kt [hunk 8]
```diff
@@ -642,6 +658,7 @@ class Cas1SpaceBookingService(
     val premisesId: UUID,
     val arrivalDate: LocalDate?,
     val departureDate: LocalDate?,
+    val characteristics: List<CharacteristicEntity> = emptyList(),
```
Comment (davidatkinsuk) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1SpaceBookingService.kt:661]: i think we should maybe not set defaults here to be consistent with other fields?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1BookingDomainEventService.kt [hunk 9]
```diff
@@ -144,6 +146,7 @@ class Cas1BookingDomainEventService(
     bookingChangedAt: OffsetDateTime,
     previousArrivalDateIfChanged: LocalDate?,
     previousDepartureDateIfChanged: LocalDate?,
+    previousCharacteristicsIfChanged: List<CharacteristicEntity>? = emptyList(),
```
Comment (davidatkinsuk) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1BookingDomainEventService.kt:149]: i'd rather the calling code explicitly stated the default value for this, if that's ok

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1BookingDomainEventService.kt [hunk 10]
```diff
@@ -524,5 +531,13 @@ class Cas1BookingDomainEventService(
...
     val isSpaceBooking: Boolean,
+    val characteristics: List<Cas1SpaceCharacteristic> = emptyList(),
+    val previousCharacteristics: List<Cas1SpaceCharacteristic>? = emptyList(),
   )
+
+  private fun List<CharacteristicEntity>.toCas1SpaceCharacteristics(): List<Cas1SpaceCharacteristic> =
+    this.map { it.asCas1SpaceCharacteristic() }
+
+  private fun CharacteristicEntity.asCas1SpaceCharacteristic() =
+    Cas1SpaceCharacteristic.entries.first { it.value == this.propertyName }
```
Comment (davidatkinsuk) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1BookingDomainEventService.kt:542]: nice

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/cas1/Cas1SpaceBookingTest.kt [hunk 11]
```diff
@@ -2077,6 +2110,45 @@ class Cas1SpaceBookingTest {
...
+      val spaceBookingBeforeUpdate = cas1SpaceBookingRepository.findByIdOrNull(spaceBooking.id)!!
+      assertThat(spaceBookingBeforeUpdate.criteria.size).isEqualTo(4)
+      assertThat(spaceBookingBeforeUpdate.criteria)
+        .extracting("modelScope", "propertyName")
+        .containsExactly(
+          tuple("room", "isArsonSuitable"),
+          tuple("room", "hasEnSuite"),
+          tuple("room", "isSingle"),
+          tuple("premises", "isWheelchairAccessible"),
+        )
```
Comment (davidatkinsuk) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/cas1/Cas1SpaceBookingTest.kt:2127]: i like the assertion before calling the endpoint, it really helps clarify the before and after values. neat!

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/Cas1SpaceBookingServiceTest.kt [hunk 12]
```diff
@@ -1929,6 +1929,7 @@ class Cas1SpaceBookingServiceTest {
           bookingChangedAt = any(),
           previousArrivalDateIfChanged = null,
           previousDepartureDateIfChanged = LocalDate.of(2025, 1, 10),
+          previousCharacteristicsIfChanged = emptyList(),
```
Comment (davidatkinsuk) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/Cas1SpaceBookingServiceTest.kt:1932]: could we maybe set this to something that isn't a default e.g. list with single value?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1BookingDomainEventService.kt [hunk 13]
```diff
@@ -249,8 +254,10 @@ class Cas1BookingDomainEventService(
             ),
             arrivalOn = bookingChangedInfo.arrivalDate,
             departureOn = bookingChangedInfo.departureDate,
+            characteristics = bookingChangedInfo.characteristics,
             previousArrivalOn = bookingChangedInfo.previousArrivalDateIfChanged,
             previousDepartureOn = bookingChangedInfo.previousDepartureDateIfChanged,
+            previousCharacteristics = bookingChangedInfo.previousCharacteristics,
```
Comment (davidatkinsuk) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1BookingDomainEventService.kt:260]: can we update the unit tests for this please - i don't think we need an additional test but we should assert that values being null for regular bookings and set to whatever is provided for space bookings?

**Review by davidatkinsuk — APPROVED** [review]
Looks good, a few comments from me

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1SpaceBookingService.kt [hunk 14]
```diff
@@ -558,6 +558,7 @@ class Cas1SpaceBookingService(
 
     val previousArrivalDate = bookingToUpdate.expectedArrivalDate
     val previousDepartureDate = bookingToUpdate.expectedDepartureDate
+    val previousCharacteristics = bookingToUpdate.criteria.toList()
```
Comment (vbala-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1SpaceBookingService.kt:561]: This creates a new immutable list. Changes to bookingToUpdate.criteria will no longer affect previousCharacteristics.

Comment (davidatkinsuk) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1SpaceBookingService.kt:561]: ohhh good spot!

**Review by davidatkinsuk — APPROVED** [review]
LGTM 👍
